panel/links.py

```diff
@@ -123,32 +172,33 @@ def init(self):
                 }
                 if not hasattr(link, 'target'):
                     pass
-                elif (type(link) is type(self) and link.source is self.source
+                elif (type(link) is type(self) and link.source is source
                     and link.target is self.target and params == link_params):
                     return
-            self.registry[self.source].append(self)
+            self.registry[source].append(self)
         else:
-            self.registry[self.source] = [self]
+            self.registry[source] = [self]
 
     @classmethod
-    def register_callback(cls, callback):
+    def register_callback(cls, callback: Type['CallbackGenerator']) -> None:
         """
         Register a LinkCallback providing the implementation for
         the Link for a particular backend.
         """
         cls._callbacks[cls] = callback
 
     @property
-    def source(self):
+    def source(self) -> Reactive | None:
```

Comment:
This is my personal preference over `Optional[T]` and what we use exclusively in bokeh, but this is 3.10 syntax, and works here only due to lazy annotations and mypy's support. See https://github.com/bokeh/bokeh/issues/11637 discussion for possible issues.

Reply:
Thanks, since bokeh was already using this I've simply adopted the decision but it'll be good to point people to that issue should it come up.